src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -400,36 +430,56 @@ private UnfilteredRowIterator applyIndexFilter(PrimaryKey key, UnfilteredRowIter
                 {
                     queryContext.rowsFiltered++;
 
-                    if (tree.isSatisfiedBy(partition.partitionKey(), (Row) unfiltered, staticRow))
+                    if (tree.isSatisfiedBy(partitionKey, (Row) unfiltered, staticRow))
                     {
-                        matchingRows.add(unfiltered);
+                        matches.add(unfiltered);
                         hasMatch = true;
+
+                        if (topK)
+                        {
+                            PrimaryKey shadowed = keyFactory.hasClusteringColumns()
+                                                  ? keyFactory.create(partitionKey, ((Row) unfiltered).clustering())
+                                                  : keyFactory.create(partitionKey);
+                            keysToShadow.remove(shadowed);
+                        }
                     }
                 }
             }
 
+            // If any non-static rows match the filter, there should be no need to shadow the static primary key:
+            if (topK && hasMatch && keyFactory.hasClusteringColumns())
+                keysToShadow.remove(keyFactory.create(partitionKey, Clustering.STATIC_CLUSTERING));
+
+            // We may not have any non-static row data to filter...
             if (!hasMatch)
             {
                 queryContext.rowsFiltered++;
 
-                if (tree.isSatisfiedBy(key.partitionKey(), staticRow, staticRow))
+                if (tree.isSatisfiedBy(partitionKey, staticRow, staticRow))
+                {
                     hasMatch = true;
+
+                    if (topK)
+                        keysToShadow.remove(keyFactory.create(partitionKey, staticRow.clustering()));
```

Comment:
If a static row matches, it means all rows in the partition match. So, should we clear `keysToShadow` here?

Reply:
I think you're right...changing...